### PR TITLE
internal low-priority: Add permission check for VO when using rule_id #19

### PR DESCRIPTION
**Pull request by patrick-austin**
internal low-priority: Add permission check for VO when using rule_id #19
------------------
See [issue #19](https://github.com/stfc/rucio/issues/19) for detailed description.

### CODE REVIEW

File: lib/rucio/tests/test_multi_vo.py [hunk 1]
```diff
@@ -47,6 +50,24 @@ def setup(self):
...
+        scope = InternalScope('mock', **self.vo)
+        dataset = 'dataset_' + str(generate_uuid())
+        account = InternalAccount('root', **self.vo)
+        rse_str = ''.join(choice(ascii_uppercase) for x in range(10))
+        rse_name = 'MOCK_%s' % rse_str
+        add_rse(rse_name, 'root', **self.vo)
+        add_did('mock', dataset, 'DATASET', 'root', **self.vo)
+        rule_id = add_rule(dids=[{'scope': scope, 'name': dataset}], account=account, copies=0, rse_expression=rse_name, grouping='NONE', weight='fakeweight', lifetime=None, locked=False, subscription_id=None)[0]
+
+        vo_api.add_vo(self.new_vo, 'super_root', 'Add new VO with super_root', 'rucio@email.com', 'def')
```
Comment (elichad) [inline lib/rucio/tests/test_multi_vo.py]: Why is the VO `new_vo` now created within each test rather than during setup? (I didn't spot this in last week's PRs, whoops)

Comment (patrick-austin) [reply lib/rucio/tests/test_multi_vo.py]: We still add a VO called 'new' during `setup()` for the client tests I wrote (which were for things like making sure lists for 'tst' don't include the stuff that's at 'new'. I didn't see any harm in always using the same VO for this as the tests shouldn't actually change the VO itself.
For the API tests, which were primarily for testing things like being able to add new VOs and that you can't add duplicates, I couldn't/didn't just use 'new' again since it already existed (and didn't want consecutive runs of tests to interfere), so randomly generated a three character string for a new VO each time. The reason why this went in the test functions and not in `setup()` is because I figured it'd be better to not always add that random VO in case not all of the core/API tests needed it - part of me was worried that because the VO is only 3 characters there was a small chance you could end up trying to add the same VO twice (and so by only adding it in the functions that needed it you reduced that likelihood).
Ironically enough a test suite for #20 failed on what looks to be that exact issue, getting the same three character string by chance twice and so failing to add the VO the second time (it never actually reached the code to assert on).
So overall this is a messy situation that would benefit from a more consistent way of adding VOs for testing purposes (maybe a separate utility function that gracefully handles duplicates?) applied across all the tests. But given that the tests are in flux right now it might be easier to do this once most of them are written, potentially creating a new issue or doing it as part of #25?

Comment (elichad) [reply lib/rucio/tests/test_multi_vo.py]: Why not just hardcode in some new third VO name for these tests?
I'm also confused as to why the `add_vo` calls in each test don't throw Duplicate exceptions after the first test in the group is run - I thought a VO by the name of `new_vo` would then exist for the rest of the tests.

Comment (patrick-austin) [reply lib/rucio/tests/test_multi_vo.py]: That would work for these tests (I think), and probably for some of the other tests like root recovery. Hardcoding the VO names for tests would stop you from being able to do the existing `add_vo` test (since it needs to be a new VO to test adding it), but given we need to successfully add a VO for the other tests to work anyway, it probably doesn't need to be explicitly tested (and the function could still test that duplication fails and that access is denied if you're not `super_root`. So yeah it probably could be done that way with a few changes if we want to use hardcoding over random strings for these kind of tests.
That also confused me, I assumed `setup()` was being called once for the class (group of tests) but it seems to actually be called once per function (so you randomly generate a three character string multiple times).

Comment (elichad) [reply lib/rucio/tests/test_multi_vo.py]: I think hardcoding here is definitely preferable to avoid test failures arising from randomisation - for 10+-character strings, random generation is fine, but it's a bit too risky for a 3-character string.
I didn't know that about `setup()` - but now I guess maybe that's the difference between having a class method for setup (e.g. as in test_account_limits.py with the `@classmethod` tag) vs a standard method. Class method might be preferable/slightly better practice, but probably worth checking that's actually how it works with nosetests

Comment (patrick-austin) [reply lib/rucio/tests/test_multi_vo.py]: I've had a look and it does seem that the `@classmethod` type setup only gets called once, so I'll change the tests to use that as well as moving towards hardcoded VO(s) in this and my other tests.
Another thing I was thinking is currently the multi_vo tests I wrote are only designed to be run in multi_vo, most if not all would fail the assertions if you ran the tests in single_vo mode. This is partially handled by excluding the file from the normal test suites, but I'm wondering if it's worth also coding it so they either don't do anything in single_vo mode or they raise an error to explicitly say that you shouldn't be running them if multi_vo is turned off. If you have any thoughts on that I'd appreciate it.

Comment (elichad) [reply lib/rucio/tests/test_multi_vo.py]: If there's a way to tell nosetests during the setup-class method to just not run the tests if it's single_vo mode, then that'd be ideal. Otherwise the class should throw at least a warning (and maybe an error) during set-up if the tests are being run in a single-VO way. Worth asking Martin as he'll probably have preferences about how to handle this (warnings vs errors vs quietly skipping tests vs letting them fail, etc)

Comment (elichad) [reply lib/rucio/tests/test_multi_vo.py]: Resolved: asked Martin, conclusion is we should log a warning and then let the tests fail.

**Review by elichad — APPROVED** [review]
